README
EtherTerm (c) 2014-2015 Michael Griffin <mrmisticismo@hotmail.com> 
An SDL2 Terminal Telnet/SSH program for Windows, Linux & OSX

Alpha Stage 1

The program is developed mainly in Windows using the (Codelite IDE) with mingw32/64 
and will compile in OSX, and Linux under GNU g++.

This is a full graphics program just like a game engine, it's not a console application 
and best runs on more recent hardware.

EtherTerm is a Telnet/SSH terminal with CP437 (ANSI) Graphics terminal emulation for 
connecting to Telnet/SSH BBS systems and shells.

Each system can be setup with a default font. Any systems using Syncterm escape 
sequences for font switching will override the default font.

Fonts currently supported are: VGA-8x16 (CP437)
-----------------------------------------------
 MicroKnightPlus-8x16 (AMIGA)
 Mo'soul-8x16 (AMIGA)
 Pot Noodle-8x16 (AMIGA)
 Topaz Plus-8x16 (AMIGA)

UTF-8 Output encoding with TTF fonts are in the works!

Text selection is almost finished. You can left click and drag the mouse around. 
To highlight text on the screen. Copying selected text will be completed soon.

Right click will paste text from the clipboard.

ALT + ENTER toggles between different window screen sizes. On started 1280x800 is 
the default resolution, The next goes Full Screen, then down to 640x400 to keep the 
proper 8x6 font resolution.


Work in progress:
-----------------
The Dialing directory is incomplete, Add, Edit, Delete, About etc.. 
Not all options are setup just yet.

To add new systems or remove, you must edit the dialingdirectory.xml 
file in a text editor located in the /assets folder.

If you want to connect to SSH servers, you must supply your username, 
and password in the dialdirectory.XML file since there are no prompts 
to enter these when you connect at the moment.

I'm also in the middle of implimenting more control sequence behavior. 
I have a lot of XTERM sequences laid out so more will be implemented with time.


Not yet implemented:
--------------------
Lots of things are still in the early testing and debugging stages. 
There are no transfer protocols at the moment.

Some keyboard keys still need to be implemented like F1 function keys. 
All others should work just fine including CTRL Keys.

There is no scroll back buffer, this is also in the works.

FTP will be added too.


To compile the following libraries must be pre-installed:
---------------------------------------------------------
SDL2
SDL2_Net
Libssh (Requires OpenSSL, Zlib)


Example Windows (mingw32/64) linking:
------------------------------------
Libs: mingw32 or 64, SDL2main, SDL2, SDL2_net, libssh


Example (G++) linking:
---------------------
libs: libSDL2, libSDL2_net, libssh


Must compile with latest c++ environment flags.

At the very least and you can also replace 0x with 11 on newer compilers. Usually g++ (4.7)+
--------------------------------------------------------------------------------------------
-std=gnu++0x or -std=c++0x


Makefile instructions:
----------------------
To build windows you must already have a compiler and the requirements. 
You can use a good IDE like CodeLite, CodeBlocks, or Visual Studio (more headache). 
You then must installs the libs and setup the include paths.

The following Makefile commands will always default for Windows, or you can specify.
------------------------------------------------------------------------------------
make
make win

make clean
make clean-win


For OSX
-------
make osx

make clean-osx


For Linux
---------
make linux

make clean-linux



Linux Build Steps:
------------------
For Ubunutu, Mint and Debian you can use APT-GET to install all requirements. Notes for Arch Linux will follow shortly once steps are completed.

Clone the Git repository.
------------------------
git clone https://github.com/M-griffin/EtherTerm.git

Install build-essential
-----------------------
sudo apt-get install build-essential

Install SDL2
------------
sudo apt-get install libsdl2-dev

Install SDL2_Net
----------------
sudo apt-get install libsdl2-net-dev

Install Libssh
--------------
sudo apt-get install libssh-dev

To compile your executable.
---------------------------
make-linux 



OSX Build Steps:
----------------

Clone the Git repository.
-------------------------
git clone https://github.com/M-griffin/EtherTerm.git

Install Brew if you don't have it already
-----------------------------------------
ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"

Install SDL2
------------
brew install sdl2

Installs /usr/local/Cellar/sdl2/2.0.3:

Install SDL2_Net
----------------
brew install sdl2_net

installs /usr/local/Cellar/sdl2_net/2.0.0

Install Libssh
--------------
brew install libssh

installs libssh /usr/local/Cellar/libssh6.4

To compile your executable.
---------------------------
make-osx


Possible issues with OSX Yosemite when building libssh. 
There is a report and confirmation of a macro redefinition error when 
compiling on Yosemite. A patch for this issue was created and applied 
just a few days ago. Earlier OSX versions are not affected.

You can use the following links for patch details, or get a new copy of 
libssh directly from the repo. Brew should have all dependencies for libssh 
already set for a clean libssh manual compile.

Bug Report
----------
https://red.libssh.org/issues/164#change-488

Patch
-----
https://red.libssh.org/attachments/download/95/patch-ConfigureChecks.cmake.diff

link to libssh repo:
--------------------
git clone http://git.libssh.org/projects/libssh.git libssh


Executing the first time:
------------------------
Once these are all set, just run the make-osx to build the executable.

EtherTerm will look for the ASSETS folder off the root executable folder

After you compile, the executable is placed in a default DEBUG folder. 
If you want to run from this folder you must copy the assets folder to 
the DEBUG folder or the root folder where the executable is located.

For Example:
------------
./EtherTerm/DEBUG   ( Contains Executable )
./EtherTerm/assets  ( Contains Fonts, Screens and xml files )

Copy the assets folder to the debug foler so it looks like this:
----------------------------------------------------------------
./EtherTerm/DEBUG/assets


Acknowledgement and Thanks:
- Dingo: Telnet options and general terminal functionality
- G00R00: Optimization ideas on scrolling pixels faster.
- (SDL Game Programming) For good ideas with Singletons and State Machines.
- Maze: Testing and bug reports
- IceDevil: Testing and bug reports


changelog.txt

Change Log 0.4.11 - 2019-08-23
- Added new use_pipe_color flag to the dialing directory.  This translates PIPE Color Codes
  that are displayed, yet not translated by older BBS Systems.  This will do the work so that 
  Message and other text on older BBS Software can look as nice as modern software without 
  ugly pipe codes displaying everywhere.

Change Log 0.4.10 - 2019-08-21
- Refactored Dialing Directory XML and FontSets to Yaml
- FontSets are regenerated automatically and 
  The Dialing Directory will detect if the XML already exists and convert
  it to the new Yaml file format.  If none exists it will generate a default entry.
- This is more an update release to get all project file updated back in sync
  There are several things that need to be rewritten or finished in the Dialing Directory
  for creating and managing new systems.  This alone with better SSH support will be 
  updated.  right now you have to put a login and password in the SystemEntry to use SSH
  Input fields and configuration are high on the list to get worked on next.
  After this more updates to transfer protocols.
  
Change Log 0.4.9 - 2019-07-11
- Fixed Issues - Most likely long standing with Posix/*nix systems and the dialing 
  directory not displaying correctly. This was fixed in 0.4.9-alpha Branch but was 
  never merged back to Master, So making a few updates until 0.4.10-alpha is ready.

Change Log 0.4.7 - 2019-07-10 - Merge to Master
- Fixed Posix issue with missing SOCKET definition on some Linux builds
- Added new PostBuild to cleanup of Makefiles to remove my develop paths to make it
  easier to build from the command line.  This will now automatically clean when i build the project
  so when code changes are committed the Makefiles will be already cleaned.
  NOTE, i don't have FreeBSD and OSX VM's working right now, so those projects will be updated lateron.
- Fixed Display bug in Posix with Dialing Directory and MCI Code Off-Sets

Change Log 0.4.7 - 2018-04-07
- Re-Added Back Protocol for testing downloads. WIP.
- Added back pulling screen size 80x25, 80x50 etc.. from Dialing Directory.

Change Log 0.4.6 - 2017-12-29
- Include SDL2_net into project, added custom fix for SO_KEEPALIVE so idle
  chat connections would not lose passed socket handles.  (fe.. MRC Chat on BBSes)
- Tested new compiles and updated all platform project files, Win32, FreeBSD, Linux, OSX,
  and ARM which uses the same Linux project but gcc7.x with Arch.
  Also include new IRC include files while the implementation is a WIP.

Change Log 0.4.5 - 2017-12-28
- ESC[C Cursor forward will no longer move to next line
- Added extra shutdown management to session manager so give sessions some time
  to finish their last updates on disconnects to avoid any exception errors.
- WIP, Started IRC Session Socket Connections and added initial irc_manager class.
  Initial connections are working, now need to parse responses and setup client
  command parsing.

Change Log 0.4.4 - 2017-12-23
- Added ESC[6n Response Queue.  Systems like Enigma BBS query 6+ ESC[6n
  sequences and without response directly to all 6, fields appears out of order.
  to correct this, a queue was setup when parsing so that all write sequences
  are compiled and sent as a single response.

Change Log 0.4.3 - 2017-12-13
- Re-Implemented ESC and menu options (Q) and Lightbar Enter for exiting program
- Re-Implemented ALT + H to hang up on connections
- Re-Implemented ALT + ETNER to toggle window sizes on connections includes full screen 
- Re-Implemented ALT + ENTER to toggle window sizes on Dialing directory minus full screen
  Full screen removed from dialing directory because it spawn each connection in it's own
  separate window, if the dialing directory is full screen it doesn't transfer focus properly.
  some more tweaks to follow.

Change Log 0.4.2 - 2017-12-11
- Created FreeBsd Project.
  Tested on FreeBSD 11 using clang LLVM
- Updated files and project for clean compile on FreeBsd and clang
- Cleaned up Linux Project files and removed left over boost references.

Change Log 0.4.1 - 2017-12-11
- Fixed Version Numbering, 0.4 prefix is more appropriate.
- Ripped out Boost Libs
- Re-Implemented new io_service after ripping out boost.
created async_read, async_write, async_connect which work as drop in drop replacements
- Re-implemented SSH from 2.x branch
- Lots of cleanup with cppCheck and cross compiling on Linux 
- WIP 3.x features rolled back, next on list to re-implement.
- WIP re-implement ALT + ENTER Window resizing to Full Screen Toggle
- WIP re-implement ALT + H To hang-up
- WIP re-implement External Protocols
- WIP re-implement Window Size (default 80x25) as dialing directory config option
- WIP re-implement ESC for exit in dialing directory

Change Log 0.4.0 - 2017-12-05
- Merged 3.x Branch into Master
- Rolled back changes to commit 3.0.42
- Re-added Committed from 3.0.45
- Added new Individual Projects for Win32 and Linux
- Several Small tweaks to line up windows and Linux builds for refactoring.
- WIP, Re-Implement 3.0.46, 3.0.44, and height/width from 3.0.43
- WIP Remove Boost and re-add SDL2_Net for Libs functionality same as 2.x branch   

Change Log 3.0.46 - 2017-10-03
- Fixed ESC[C Cursor Movement to not wrap when it hits end of Term Width.

Change Log 3.0.45 - 2016-09-04
- Added Some Extra Checks in Telnet Options to remove responses for 
For invalid arguments, to avoid option looks.

Change Log 3.0.44 - 2016-05-19
- Added support for external protocols for file transfers.
- Initial tests with Zmodem download are working
- Probably add options and create protocol.xml file for x, y, z modem
  and the command lines for each.

Change Log 3.0.43 - 2016-05-09
- Added Height/Width settings Terminal settings to dialing directory xml
- Added Texture Filtering settings to the dialing directory xml
- Note, that current window setup is to resize the screen to the window
  This just happens ignore texture filtering. So part II needs to 

telnet.hpp
#ifndef TELNET_HPP
#define TELNET_HPP

// Temp
#define NUM_LINES 25

//Definitions for the TELNET protocol.
#define    IAC     255        // interpret as command:
#define    DONT    254        // you are not to use option
#define    DO      253        // please, you use option
#define    WONT    252        // I won't use option
#define    WILL    251        // I will use option
#define    SB      250        // interpret as sub-negotiation
#define    GA      249        // you may reverse the line
#define    EL      248        // erase the current line
#define    EC      247        // erase the current character
#define    AYT     246        // are you there
#define    AO      245        // abort output--but let prog finish
#define    IP      244        // interrupt process--permanently
#define    BREAK   243        // break
#define    DM      242        // data mark--for connect. cleaning
#define    NOP     241        // nop
#define    SE      240        // end sub negotiation
#define    EOR     239        // end of record (transparent mode)
#define    ABORT   238        // Abort process
#define    SUSP    237        // Suspend process
#define    xEOF    236        // End of file: EOF is already used...
#define    SYNCH   242        // for telfunc calls

#ifdef TELCMDS
const char *telcmds[] =
{
    "EOF", "SUSP", "ABORT", "EOR",
    "SE", "NOP", "DMARK", "BRK", "IP", "AO", "AYT", "EC",
    "EL", "GA", "SB", "WILL", "WONT", "DO", "DONT", "IAC", 0
};
#else
extern const char *telcmds[];
#endif

#define    TELCMD_FIRST  xEOF
#define    TELCMD_LAST   IAC
#define    TELCMD_OK(x)  ((unsigned int)(x) <= TELCMD_LAST && \
                          (unsigned int)(x) >= TELCMD_FIRST)

#define    TELCMD(x)      telcmds[(x)-TELCMD_FIRST]

// telnet options
#define TELOPT_BINARY          0    // 8-bit data path
#define TELOPT_ECHO            1    // echo
#define TELOPT_RCP             2    // prepare to reconnect
#define TELOPT_SGA             3    // suppress go ahead
#define TELOPT_NAMS            4    // approximate message size
#define TELOPT_STATUS          5    // give status
#define TELOPT_TM              6    // timing mark
#define TELOPT_RCTE            7    // remote controlled transmission and echo
#define TELOPT_NAOL            8    // negotiate about output line width
#define TELOPT_NAOP            9    // negotiate about output page size
#define TELOPT_NAOCRD         10    // negotiate about CR disposition
#define TELOPT_NAOHTS         11    // negotiate about horizontal tab-stops
#define TELOPT_NAOHTD         12    // negotiate about horizontal tab disposition
#define TELOPT_NAOFFD         13    // negotiate about form-feed disposition
#define TELOPT_NAOVTS         14    // negotiate about vertical tab stops
#define TELOPT_NAOVTD         15    // negotiate about vertical tab disposition
#define TELOPT_NAOLFD         16    // negotiate about output LF disposition
#define TELOPT_XASCII         17    // extended ASCII character set
#define TELOPT_LOGOUT         18    // force log out
#define TELOPT_BM             19    // byte macro
#define TELOPT_DET            20    // data entry terminal
#define TELOPT_SUPDUP         21    // supdup protocol
#define TELOPT_SUPDUPOUTPUT   22    // supdup output
#define TELOPT_SNDLOC         23    // send location
#define TELOPT_TTYPE          24    // terminal type
#define TELOPT_EOR            25    // end or record
#define TELOPT_TUID           26    // TACACS user identification
#define TELOPT_OUTMRK         27    // output marking
#define TELOPT_TTYLOC         28    // terminal location number
#define TELOPT_3270REGIME     29    // 3270 regime
#define TELOPT_X3PAD          30    // X.3 PAD
#define TELOPT_NAWS           31    // window size
#define TELOPT_TSPEED         32    // terminal speed
#define TELOPT_LFLOW          33    // remote flow control
#define TELOPT_LINEMODE       34    // Line-mode option
#define TELOPT_XDISPLOC       35    // X Display Location
#define TELOPT_OLD_ENVIRON    36    // Old - Environment variables
#define TELOPT_AUTHENTICATION 37    // Authenticate
#define TELOPT_ENCRYPT        38    // Encryption option
#define TELOPT_NEW_ENVIRON    39    // New - Environment variables
#define TELOPT_TN3270E        40    // RFC2355 - TN3270 Enhancements
#define TELOPT_CHARSET        42    // RFC2066 - Charset
#define TELOPT_COMPORT        44    // RFC2217 - Com Port Control
#define TELOPT_KERMIT         47    // RFC2840 - Kermit
#define TELOPT_EXOPL         255    // extended-options-list

// MUD WIP
#define TELOPT_MSDP           69    // used to send mud server data
#define TELOPT_MSSP           70    // used to send mud server information
#define TELOPT_MCCP_1         85    // MUD zlib stream compression Version 1
#define TELOPT_MCCP_2         86    // MUD zlib stream compression Version 2 RFC1950.
#define TELOPT_MSP            90    // used to toggle Mud Sound Protocol
#define TELOPT_MXP            91    // used to toggle Mud Extension Protocol

#define TELOPT_FIRST    TELOPT_BINARY
#define TELOPT_LAST     TELOPT_KERMIT

#define TELOPT_OK(x)    ((unsigned int)(x) <= TELOPT_LAST)
#define TELOPT(x)       telopts[(x)-TELOPT_FIRST]

#define NTELOPTS        (1+TELOPT_LAST)

#ifdef TELOPTS
const char *telopts[NTELOPTS+1] =
{
    "BINARY", "ECHO", "RCP", "SUPPRESS GO AHEAD", "NAME",
    "STATUS", "TIMING MARK", "RCTE", "NAOL", "NAOP",
    "NAOCRD", "NAOHTS", "NAOHTD", "NAOFFD", "NAOVTS",
    "NAOVTD", "NAOLFD", "EXTEND ASCII", "LOGOUT", "BYTE MACRO",
    "DATA ENTRY TERMINAL", "SUPDUP", "SUPDUP OUTPUT",
    "SEND LOCATION", "TERMINAL TYPE", "END OF RECORD",
    "TACACS UID", "OUTPUT MARKING", "TTYLOC",
    "3270 REGIME", "X.3 PAD", "NAWS", "TSPEED", "LFLOW",
    "LINEMODE", "XDISPLOC", "OLD-ENVIRON", "AUTHENTICATION",
    "ENCRYPT", "NEW-ENVIRON", "TN3270E", "CHARSET", "COM-PORT",
    "KERMIT",
    0
};
#endif

// sub-option qualifiers
#define    TELQUAL_IS          0    // option is...
#define    TELQUAL_SEND        1    // send option
#define    TELQUAL_INFO        2    // ENVIRON: informational version of IS
#define    TELQUAL_REPLY       2    // AUTHENTICATION: client version of IS
#define    TELQUAL_NAME        3    // AUTHENTICATION: client version of IS

#define    LFLOW_OFF           0    // Disable remote flow control
#define    LFLOW_ON            1    // Enable remote flow control
#define    LFLOW_RESTART_ANY   2    // Restart output on any char
#define    LFLOW_RESTART_XON   3    // Restart output only on XON

// LINE-MODE sub-options
#define    LM_MODE            1
#define    LM_FORWARDMASK     2
#define    LM_SLC             3

#define    MODE_EDIT        0x01
#define    MODE_TRAPSIG     0x02
#define    MODE_ACK         0x04
#define    MODE_SOFT_TAB    0x08
#define    MODE_LIT_ECHO    0x10

#define    MODE_MASK        0x1f

// Not part of protocol, but needed to simplify things...
#define MODE_FLOW        0x0100
#define MODE_ECHO        0x0200
#define MODE_INBIN       0x0400
#define MODE_OUTBIN      0x0800
#define MODE_FORCE       0x1000

#define    SLC_SYNCH   1
#define    SLC_BRK     2
#define    SLC_IP      3
#define    SLC_AO      4
#define    SLC_AYT     5
#define    SLC_EOR     6
#define    SLC_ABORT   7
#define    SLC_EOF     8
#define    SLC_SUSP    9
#define    SLC_EC      10
#define    SLC_EL      11
#define    SLC_EW      12
#define    SLC_RP      13
#define    SLC_LNEXT   14
#define    SLC_XON     15
#define    SLC_XOFF    16
#define    SLC_FORW1   17
#define    SLC_FORW2   18
#define    SLC_MCL     19
#define    SLC_MCR     20
#define    SLC_MCWL    21
#define    SLC_MCWR    22
#define    SLC_MCBOL   23
#define    SLC_MCEOL   24
#define    SLC_INSRT   25
#define    SLC_OVER    26
#define    SLC_ECR     27
#define    SLC_EWR     28
#define    SLC_EBOL    29
#define    SLC_EEOL    30

#define    NSLC        30

// For backwards compatibility, we define SLC_NAMES to be the
// list of names if SLC_NAMES is not defined.

#define    SLC_NAMELIST    "0", "SYNCH", "BRK", "IP", "AO", "AYT", "EOR",    \
    "ABORT", "EOF", "SUSP", "EC", "EL", "EW", "RP",    \
    "LNEXT", "XON", "XOFF", "FORW1", "FORW2",    \
    "MCL", "MCR", "MCWL", "MCWR", "MCBOL",        \
    "MCEOL", "INSRT", "OVER", "ECR", "EWR",        \
    "EBOL", "EEOL",                    \
    0

#ifdef    SLC_NAMES
char *slc_names[] =
{
    SLC_NAMELIST
};
#else
extern char *slc_names[];
#define    SLC_NAMES SLC_NAMELIST
#endif

#define    SLC_NAME_OK(x)    ((unsigned int)(x) <= NSLC)
#define SLC_NAME(x)    slc_names[x]

#define    SLC_NOSUPPORT    0
#define    SLC_CANTCHANGE   1
#define    SLC_VARIABLE     2
#define    SLC_DEFAULT      3
#define    SLC_LEVELBITS    0x03

#define    SLC_FUNC         0
#define    SLC_FLAGS        1
#define    SLC_VALUE        2

#define    SLC_ACK          0x80
#define    SLC_FLUSHIN      0x40
#define    SLC_FLUSHOUT     0x20

#define    OLD_ENV_VAR      1
#define    OLD_ENV_VALUE    0
#define    NEW_ENV_VAR      0
#define    NEW_ENV_VALUE    1
#define    ENV_ESC          2
#define ENV_USERVAR         3

//AUTHENTICATION suboptions


//Who is authenticating who ...
#define    AUTH_WHO_CLIENT      0    // Client authenticating server
#define    AUTH_WHO_SERVER      1    // Server authenticating client
#define    AUTH_WHO_MASK        1

// amount of authentication done

#define    AUTH_HOW_ONE_WAY    0
#define    AUTH_HOW_MUTUAL     2
#define    AUTH_HOW_MASK       2

// should we be encrypting? (not yet formally standardized)

#define AUTH_ENCRYPT_OFF        0
#define AUTH_ENCRYPT_ON         4
#define AUTH_ENCRYPT_MASK       4

#define    AUTHTYPE_NULL           0
#define    AUTHTYPE_KERBEROS_V4    1
#define    AUTHTYPE_KERBEROS_V5    2
#define    AUTHTYPE_SPX            3
#define    AUTHTYPE_MINK           4
#define    AUTHTYPE_SRA            6
#define    AUTHTYPE_CNT            7

#define    AUTHTYPE_TEST          99

#ifdef    AUTH_NAMES
const char *authtype_names[] =
{
    "NULL", "KERBEROS_V4", "KERBEROS_V5", "SPX", "MINK", nullptr, "SRA",
    0
};
#else
extern const char *authtype_names[];
#endif

#define    AUTHTYPE_NAME_OK(x)    ((unsigned int)(x) < AUTHTYPE_CNT)
#define    AUTHTYPE_NAME(x)    authtype_names[x]

// ENCRYPTION sub-options
#define    ENCRYPT_IS           0    // I pick encryption type ...
#define    ENCRYPT_SUPPORT      1    // I support encryption types ...
#define    ENCRYPT_REPLY        2    // Initial setup response
#define    ENCRYPT_START        3    // Am starting to send encrypted
#define    ENCRYPT_END          4    // Am ending encrypted
#define    ENCRYPT_REQSTART     5    // Request you start encrypting
#define    ENCRYPT_REQEND       6    // Request you send encrypting
#define    ENCRYPT_ENC_KEYID    7
#define    ENCRYPT_DEC_KEYID    8
#define    ENCRYPT_CNT          9

#define    ENCTYPE_ANY          0
#define    ENCTYPE_DES_CFB64    1
#define    ENCTYPE_DES_OFB64    2
#define    ENCTYPE_CNT          3

#ifdef    ENCRYPT_NAMES
const char *encrypt_names[] =
{
    "IS", "SUPPORT", "REPLY", "START", "END",
    "REQUEST-START", "REQUEST-END", "ENC-KEYID", "DEC-KEYID",
    0
};
const char *enctype_names[] =
{
    "ANY", "DES_CFB64",  "DES_OFB64",
    0
};
#else
extern const char *encrypt_names[];
extern const char *enctype_names[];
#endif

#define    ENCRYPT_NAME_OK(x)    ((unsigned int)(x) < ENCRYPT_CNT)
#define    ENCRYPT_NAME(x)        encrypt_names[x]

#define    ENCTYPE_NAME_OK(x)    ((unsigned int)(x) < ENCTYPE_CNT)
#define    ENCTYPE_NAME(x)        enctype_names[x]

#endif // TELNET_HPP
